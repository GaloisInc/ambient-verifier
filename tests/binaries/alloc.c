#include <stdlib.h>
#include <unistd.h>
#include <sys/syscall.h>

#define CALLOC_NUM 5

/**
 * This test verifies that the slab allocator places newly allocated memory
 * directly before previously allocated memory by writing off the end of one
 * allocation and into another, which alters control flow into an `execve`
 */

int main(void) {
  int* p1 = malloc(sizeof(int));
  int* p2 = calloc(CALLOC_NUM, sizeof(int));

  *p1 = 0;
  for (int i = 0; i <= CALLOC_NUM; ++i) {
    // write off end of p2 into p1
    p2[i] = i;
  }

  if (*p1) {
    syscall(SYS_execve, "/bin/sh", NULL, NULL);
  }

  return 0;
}

void _start(void) {
  main();
}
