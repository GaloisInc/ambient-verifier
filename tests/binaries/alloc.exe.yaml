statechart:
  name: BumpAlloc
  root state:
    name: Root
    initial: NormalExecution
    states:
      - name: NormalExecution
        description: |
          Execution before the out of bounds write
        transitions:
          - target: DidExec
            event: IssuesExecveSyscall
      - name: DidExec
        description: |
          Control flow compromised
        type: final
