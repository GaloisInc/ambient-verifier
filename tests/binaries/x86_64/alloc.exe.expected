successful: 0
failed: 0
wmTargets: []
